src/context.rs

```diff
@@ -468,6 +398,32 @@ impl<A: Actor> Context<A> {
 
         Ok(fut)
     }
+
+    fn inbox(&self) -> impl FusedStream<Item = Either<BroadcastMessage<A>, AddressMessage<A>>> {
+        let addr_rx = self.receiver.clone();
+        let broadcast_rx = self.broadcast_receiver.clone();
+
+        let broadcast_stream =
+            futures_util::stream::unfold(broadcast_rx, |broadcast_rx| async move {
+                match broadcast_rx.recv_async().await {
+                    Ok(msg) => Some((msg, broadcast_rx)),
+                    Err(_) => None,
+                }
+            })
+            .map(Either::Left)
+            .boxed();
+
+        let address_stream = addr_rx.into_stream().map(Either::Right).boxed();
```

Comment:
This is unfortunately a very bad idea since streams have very unfavourable performance characteristics; I can elaborate on it a bit later

Reply:
Ok, here is the problem with streams as I understand it. Most of this comes with a conversation I had with zesterer (the main developer of flume) and stjepang about streams and channels.

With a future-based receive, you as the channel writer have a way to *roughly*[^1] de-register interest in receiving. How channels such as flume works is that when you call `recv`, if there isn't a message, it will put the waker into a queue of receivers interested in receiving a message. Then, when a message is sent, one is popped off, the message is delivered to it, and the waker is woken. When a future is dropped, their waker is removed from the queue, regardless of if it has received a message. So, to summarise, exactly one receiver is woken to take exactly one message.

However, with a stream, it is added to a separate queue, and streams are woken for *every message*, since there is no way to know if a stream is still interested, as you don't control the stream's receive future implementation. This creates an unavoidable thundering herd problem. I'm not aware of any way to work around this - flume has this property, as does async-channel.

[^1]: the Drop implementation of the receive future is usually used to see when a receiver is de-registering interest. You *can* `mem::forget` a future but the general consensus is that if you do this you void your guarantees on receiving all messages.

Reply:
Thanks for the thorough explanation!

If I understand correctly, we should be fine though if we base both streams on top of `unfold` and the `async_recv` function? That way, only the immediate next item is registered and if we drop the stream, the future will be dropped as well.

Reply:
Also, would you be open to the idea of completely encapsulating this API so the streams are an implementation detail and merge something like this despite non-ideal performance characteristics?

Once they are an implementation detail, we can iterate on internal designs to improve performance and have the API out there.

Reply:
Maybe, it depends exactly how its done. Would you mind writing it out or pushing what you have in mind exactly?

Reply:
> Maybe, it depends exactly how its done. Would you mind writing it out or pushing what you have in mind exactly?

See https://github.com/Restioson/xtra/pull/51/commits/d76c2f82132a25df7c76d411681ae7de59df1d21.